File: go/inorder_traversal.go
package algos

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func help(root *TreeNode, result []int) []int {
	if root != nil {
		result = help(root.Left, result)
		result = append(result, root.Val)
		result = help(root.Right, result)
	}
	return result
}

func inorderTraversal(root *TreeNode) []int {
	var result []int
	return help(root, result)
}

File: rust/src/algos/inorder_traversal.rs
use std::rc::Rc;
use std::cell::RefCell;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}


pub fn help(root: &Option<Rc<RefCell<TreeNode>>>, result: &mut Vec<i32>) {
    if let Some(node) = root {
        let node = node.borrow();
        help(&node.left, result);
        result.push(node.val);
        help(&node.right, result);
    }
}

pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let mut result: Vec<i32> = Vec::new();
    help(&root, &mut result);
    result
}

